src/CodeGen_Vulkan_Dev.cpp

```diff
@@ -0,0 +1,2924 @@
+            {"sqrt_f32", GLSLstd450Sqrt},
+            {"tan_f16", GLSLstd450Tan},
+            {"tan_f32", GLSLstd450Tan},
+            {"tanh_f16", GLSLstd450Tanh},
+            {"tanh_f32", GLSLstd450Tanh},
+            {"trunc_f16", GLSLstd450Trunc},
+            {"trunc_f32", GLSLstd450Trunc},
+            {"mix", GLSLstd450FMix},
+        };
+
+        // The SPIRV-IR builder
+        SpvBuilder builder;
+
+        // The scope contains both the symbol id and its storage class
+        using SymbolIdStorageClassPair = std::pair<SpvId, SpvStorageClass>;
+        using SymbolScope = Scope<SymbolIdStorageClassPair>;
+        using ScopedSymbolBinding = ScopedBinding<SymbolIdStorageClassPair>;
+        SymbolScope symbol_table;
+
+        // Map from a variable ID to its corresponding storage type definition
+        struct StorageAccess {
+            SpvStorageClass storage_class = SpvStorageClassMax;
+            uint32_t storage_array_size = 0;  // zero if not an array
+            SpvId storage_type_id = SpvInvalidId;
+            Type storage_type;
+        };
+        using StorageAccessMap = std::unordered_map<SpvId, StorageAccess>;
+        StorageAccessMap storage_access_map;
+
+        // Defines the binding information for a specialization constant
+        // that is exported by the module and can be overriden at runtime
+        struct SpecializationBinding {
+            SpvId constant_id = 0;
+            uint32_t type_size = 0;
+            std::string constant_name;
+        };
+        using SpecializationConstants = std::vector<SpecializationBinding>;
+
+        // Defines a shared memory allocation
+        struct SharedMemoryAllocation {
+            SpvId constant_id = 0;  // specialization constant to dynamically adjust array size (zero if not used)
+            uint32_t array_size = 0;
+            uint32_t type_size = 0;
+            std::string variable_name;
+        };
+        using SharedMemoryUsage = std::vector<SharedMemoryAllocation>;
+
+        // Defines the specialization constants used for dynamically overiding the dispatch size
+        struct WorkgroupSizeBinding {
+            SpvId local_size_constant_id[3] = {0, 0, 0};  // zero if unused
+        };
+
+        // Keep track of the descriptor sets so we can add a sidecar to the
+        // module indicating which descriptor set to use for each entry point
+        struct DescriptorSet {
+            std::string entry_point_name;
+            uint32_t uniform_buffer_count = 0;
+            uint32_t storage_buffer_count = 0;
+            SpecializationConstants specialization_constants;
+            SharedMemoryUsage shared_memory_usage;
+            WorkgroupSizeBinding workgroup_size_binding;
+        };
+        using DescriptorSetTable = std::vector<DescriptorSet>;
+        DescriptorSetTable descriptor_set_table;
+
+        // The workgroup size.  May vary between kernels.
+        uint32_t workgroup_size[3];
```

Comment:
Why 3? As a non-GPU person this seems like a random magic number.

Reply:
Ah, yes ... workgroup size is the extents for the three dimensional index space used as part of the kernel dispatch. It can also be used to adjust the layout for work items (aka GPU threads), based on logical groupings.

Reply:
Ill add a comment clarifying its usage! Thx!